### PR TITLE
Basic functional testing

### PR DESCRIPTION
**Pull request by ewandennis**
### What Changed
 - Pulled the route rendering logic out of `App` into `AppRoutes` for easier testing
 - Created a simple helper for mounting a cut down app at a named route for functional testing
    - Uses [Enzyme's full DOM rendering](https://airbnb.io/enzyme/docs/api/mount.html) (aka `mount()`)
    - Added portal setup steps to `config/jest/setup.js` to support testing
 - Added small functional tests to cover 2 recent problem areas and showcase the approach
   -  A/B testing
   - Events search
 - Refactored integration testing helpers for easier use in func tests if needed

### How To Test
 - `npm test` will run unit, integration and functional tests
 - Func tests are held in `src/__func__`

**Consider reviewing commits separately.**

### To Do
- [ ] Document the approach
- [ ] Provide guidance on when to use each type of testing
- [ ] Address feedback
- [ ] Move `src/__integration__/http-responses` to `src/__testHelpers__` once review is complete.

### CODE REVIEW

**Comment by ewandennis** [comment]
For the interested, I have the following guidance in mind for when to apply each class of test:

 - write unit tests for all reusable objects to document and verify their interfaces
 - consider writing a storybook for reusable React components to outline use cases
 - write integration tests for complex API interactions to verify request sequencing and payloads
 - write functional tests for top-level and high value features to verify key flows

File: config/jest/setup.js [hunk 1]
```diff
@@ -30,6 +31,8 @@ jest.mock('src/components/notifications/staticMarkdownNotifications', () => [
   }
 ]);
 
+setupPortals();
```
Comment (bkemper) [inline config/jest/setup.js]: Ugh, I wish the default `testEnvironment`, `jest-environment-jsdom`, allowed us to configure a default template. 
 I don't see a way to configure it with `testEnvironmentOptions`.  The [template appears to be static](https://github.com/facebook/jest/blob/3e32813b1bf8c91dd59814322aa9552e7c0ab2a8/packages/jest-environment-jsdom/src/index.ts#L40).

File: src/App.js [hunk 2]
```diff
@@ -34,25 +30,7 @@ const App = () => (
...
-                        <Redirect to={{ ...location, pathname: route.redirect }} />
-                      )} />
-                    );
-                  }
-
-                  return <MyRoute key={route.path} {...route} />;
-                })
-              }
-            </Switch>
+            <AppRoutes />
```
Comment (bkemper) [inline src/App.js]: 👍

File: src/AppRoutes.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import {
+  Redirect,
+  Route,
+  Switch
+} from 'react-router-dom';
+import { PublicRoute, ProtectedRoute } from 'src/components/auth';
+import routes from 'src/config/routes';
+
+export default () => (
```
Comment (bkemper) [inline src/AppRoutes.js]: Can we name this component, `const AppRoutes  = () => (` and export at the bottom?

File: src/__func__/events.test.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import mountRoute from 'src/__testHelpers__/mountRoute';
+
+Date.now = jest.fn(() => 1512509841582);
```
Comment (bkemper) [inline src/__func__/events.test.js]: Should we freeze time for all tests?

File: src/__func__/events.test.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+test('Events search page', async () => {
+  const { mounted } = await mountRoute('/reports/message-events');
+  expect(mounted.find('TableCollection')).toExist();
+});
+
+test('Events search pagination', async () => {
+  const { mounted, forceUpdate } = await mountRoute('/reports/message-events');
+  const nextBtn = mounted.find('Pager Button[disabled=false]');
+  nextBtn.simulate('click');
+  await forceUpdate();
```
Comment (bkemper) [inline src/__func__/events.test.js]: Nice to have, build this wait into a `simulate` fn.  Something like `simulate('Pager Button[disabled=false]', 'click')`.

File: src/__testHelpers__/mountRoute.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+import configureStore from 'src/store';
+
+import AppRoutes from 'src/AppRoutes';
+import { authenticate } from 'src/actions/auth';
+import { initializeAccessControl } from 'src/actions/accessControl';
+
+import asyncFlush from 'src/__testHelpers__/asyncFlush';
+
+export default async function renderRoute(route, { authenticated = true } = {}) {
```
Comment (bkemper) [inline src/__testHelpers__/mountRoute.js]: `renderRoute` ~> `mountRoute`

File: src/AppRoutes.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+              <Redirect to={{ ...location, pathname: route.redirect }} />
+            )}
+          />
+        );
+      }
+
+      return <MyRoute key={route.path} {...route} />;
+    })}
+  </Switch>
+);
```
Comment (bkemper) [inline src/AppRoutes.js]: Should this be in `src/` or `src/components`?

File: src/__testHelpers__/mountRoute.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  if (authenticated) {
+    await store.dispatch(authenticate('test-username', 'password'));
+    await store.dispatch(initializeAccessControl());
+  }
+
+  const mounted = mount(
+    <Provider store={store}>
+      <MemoryRouter initialEntries={[route]}>
+        <AppRoutes />
```
Comment (bkemper) [inline src/__testHelpers__/mountRoute.js]: Is this enough or should we include some of the other pieces, `ErrorBoundary`, `Poll`, `AuthenticationGate`, `Layout`, `Support`, `GlobalAlertWrapper`, etc.

File: src/__testHelpers__/mountRoute.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      </MemoryRouter>
+    </Provider>
+  );
+
+  await asyncFlush();
+  mounted.update();
+
+  return {
+    mounted,
```
Comment (bkemper) [inline src/__testHelpers__/mountRoute.js]: Can we call this `wrapper`?

File: src/__testHelpers__/mountRoute.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    forceUpdate: async () => {
+      await asyncFlush();
+      mounted.update();
+    },
+    store,
```
Comment (bkemper) [inline src/__testHelpers__/mountRoute.js]: Should we expose store or force users to use the `wrapper`?

File: src/__testHelpers__/setupPortals.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Our Modal and Alert components render into portals. To allow use of these components
+ * in a jsdom-driven test environment, we must manually create portals for them ahead of
+ * time.
+ */
+
+export function mkPortal(id) {
+  const node = global.document.createElement('div');
+  node.setAttribute('id', id);
+  global.document.body.appendChild(node);
```
Comment (bkemper) [inline src/__testHelpers__/setupPortals.js]: Will work if we don't [`attachTo`](https://airbnb.io/enzyme/docs/api/mount.html#arguments) jsdom?

Comment (bkemper) [reply src/__testHelpers__/setupPortals.js]: Maybe we could do this all on mount...

```
document.body.innerHTML = `
  <div>
    <div id="react-app" />
    <div id="alert-portal" />
    <div id="modal-portal" />
  </div>
`;

const mounted = mount(
  <Provider store={store}>
    <MemoryRouter initialEntries={[route]}>
      <AppRoutes />
    </MemoryRouter>
  </Provider>,
  {  attachTo: document.getElementById('react-app') }
);
```

File: src/__func__/events.test.js (hunk 4 above)
Comment (ewandennis) [reply src/__func__/events.test.js]: Both embarrassing and unnecessary.

File: src/__testHelpers__/mountRoute.js (hunk 10 above)
Comment (ewandennis) [reply src/__testHelpers__/mountRoute.js]: I swithered a little but you're right, the store is an implementation detail.

File: src/__testHelpers__/mountRoute.js (hunk 8 above)
Comment (ewandennis) [reply src/__testHelpers__/mountRoute.js]: Yeah, I couldn't settle now how much to bring into this environment. We could just use `App` and stub out the annoying bits like analytics. That might bring some more complex initialisation with artificial wait state and suchlike. Will give it some thought.

File: src/__testHelpers__/setupPortals.js (hunk 11 above)
Comment (ewandennis) [reply src/__testHelpers__/setupPortals.js]: There's a timing issue here in that our `innerHtml` is not parsed/applied/DOM not updated by the time our test runs.

I'm sure there are more isolated solutions. This one has the benefit of working and being relatively intuitive.

File: src/__testHelpers__/mountRoute.js (hunk 8 above)
Comment (bkemper) [reply src/__testHelpers__/mountRoute.js]: Most of those services are guarded by configuration.  Lets just configure our test environment without them.

**Pull request merged**